### DISCUSSION TITLE
실행 컨텍스트 질문 목록

### DISCUSSION THREAD

**Discussion by lledellebell** [diving-javascript/you-dont-know-js-yet#18] [실행 컨텍스트(Execution Context)]
## 질문 목록
- #21  
- #22 
- #23 
- #24
- #25
- #26
- #27 
- #28
- #29

**Comment by wogus216** [comment]
#21 
자바스크립트에서 실행 컨텍스트는 코드가 실행되는 환경을 정의하고, 코드 실행 중에 발생하는 다양한 정보를 관리하는 매우 중요한 개념입니다. 이는 자바스크립트 엔진이 코드를 처리하는 방식을 이해하는 데 핵심적인 역할을 합니다. 실행 컨텍스트는 코드에 대한 정보를 캡슐화하고, 각 코드 블록이나 함수가 실행될 때 필요한 모든 정보를 제공합니다

#22 
 자바스크립트 엔진은 실행 컨텍스트를 다룰 때 여러 단계로 나누어 작업을 진행합니다. 이러한 단계적 접근은 코드를 효율적으로 처리하고, 각 코드 블록이 필요로 하는 환경을 정확하게 설정하기 위해 필수적입니다. 

생성 단계 (Creation Phase): 이 단계에서 자바스크립트 엔진은 스코프, 변수, 함수 선언 등이 담긴 변수 객체를 생성하고, 스코프 체인을 형성하며, this 키워드의 값을 결정합니다. 이는 코드 실행 전에 실행 컨텍스트를 준비하는 초기 설정 단계라고 할 수 있습니다.
활성화 단계 (Activation / Code Execution Phase): 실제 코드가 실행되는 단계입니다. 변수 인스턴스화와 함수 선언의 메모리 할당이 완료된 후, 코드 내의 각종 연산이나 함수 호출 등이 실행됩니다. 이 단계에서 변수에 값이 할당되고, 실제 연산이 처리됩니다.
소멸 단계 (Teardown Phase): 함수 호출이 완료되고 해당 실행 컨텍스트가 필요 없어지면, 자바스크립트 엔진은 해당 컨텍스트를 스택에서 제거하고 관련 자원을 정리합니다. 이는 메모리를 효율적으로 관리하고, 불필요한 자원 사용을 줄이기 위한 단계입니다.

#23 
먼저 호이스팅이란 변수선언이 해당 스코프의 최상단으로 끌어올려진것 처럼 동작하는 JavaScript의 행동

1.var로 선언된 변수는 초기화와 함께 호이스팅됩니다. 이는 변수가 선언된 위치와 관계없이 함수의 시작부분에 메모리에 할당된 다는 것을 의미합니다.
그러나 실제로 할당은 코드에 도달했을 때 이루어짐. 따라서 변수를 선언하기 전에 접근하면 undefined가 반환됨
2. let 과 const  로 선언된 변수는 호이스팅되지만 초기화는 되지 않습니다. 이를 일시적 사각지대(TDZ)라고 하며, 변수 선언 전에 해당 변수를 참조하려고 하면 참조에러가 발생

스코프 규칙

1. var은 함수 스코프를 가집니다. 즉, var 변수는 선언된 함수 내에서 어디서든지 접근할 수 있으며,함수 밖에서는 접근 할 수 없음.
2. let과 const는 블록 스코프를 가집니다. 블록 외부에서는 접근할 수 없습니다.


#25 
this는 실행 컨텍스트가 생성될 때 결정된다.
this 값은 주로 다음 네 가지 방법

1. 일반함수: 함수가 일반적으로 호출되었기 때문에 this는 글로벌 객체를 가리킵니다. 참고, . 엄격 모드('use strict')에서는 this가 undefined로 설정됩니다.
2. 메소드: 객체의 메소드로서 함수가 호출될 때, this는 그 메소드를 호출한 객체를 가리킵니다.
3. 화살표 함수: this는 항상 상위 스코프의 this를 가리킵니다. 화살표 함수 자체는 this에 대한 자체 바인딩을 가지지 않습니다.
4. 브라우저에서 DOM 요소에 이벤트 핸들러로 함수를 지정할 때, this는 이벤트가 발생한 요소를 가리킵니다.

this 바인딩의 특수 케이스
화살표 함수: 화살표 함수는 자신의 this를 가지지 않습니다. 대신, 화살표 함수는 자신을 포함하고 있는 외부 함수의 this를 "상속"받습니다. 이것은 렉시컬 스코핑(lexical scoping)이라고 하며, 화살표 함수가 정의된 위치에 따라 this가 결정됩니다.
bind/call/apply 메소드: 이 메소드들은 함수의 this 값을 명시적으로 설정할 수 있습니다. bind는 새로운 함수를 반환하며, call과 apply는 즉시 함수를 호출합니다. 이 메소드들을 사용하면 개발자는 함수 호출 시 this의 값이 무엇인지 정확히 지정할 수 있습니다.

#29 
실행컨텍스트는 메모리 할당을 관리하는 구조이며, 이는 가비지 컬렉션의 대상
가비지 컬렉션은 프로그램이 더 이상 사용하지 않는 메모리를 자동으로 회수하는 과정

**실행 컨텍스트 생성과 소멸**
각 함수 호출 시 새로운 실행 컨텍스트가 스택에 푸시되며, 함수 실행이 완료되면 해당 컨텍스트는 스택에서 팝됩니다.
실행 컨텍스트가 스택에서 제거되면 그 안에 포함된 모든 지역 변수와 함수 선언 등도 메모리에서 해제될 수 있습니다. 단, 클로저가 있을 경우 해당 변수는 여전히 접근 가능하며 메모리에 남아 있게 됩니다.


**메모리 관리 최적화 전략**
1. 변수의 적절한 생명주기 관리: 필요한 최소 범위에서만 선언하고 사용
2. 클로저 사용 최소화: 사용하지 않는 데이터를 메모리에 게속 유지하기에 신중하게 사용 클로저가 필요없어지면 null이나 nudefined 로 설정하여 참조를 명시적으로 끊어주는 것이 좋음
3. 이벤트 리스너 제거: 이벤트 리스너는 DOM요소와 같이 큰 객체를 참조 할 수 있으며, 이는 메모리의 누수의 원인이 될 수 있음
4. 대규모 데이터 구조 관리: 큰 배열이나 겍체등은 사용 후 적절히 정리하거나 필요한 부분만 추출하여 저장


실행 컨텍스트와 메모리 관리를 최적화하면 애플리케이션의 성능과 반응성을 크게 향상 가능
